framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,656 @@
+     * []}, then {@code @Baz []}, and then finally {@code @Foo int}. This implementation therefore
+     * passes a string builder with the result of the array types seen so far, to handle
+     * multidimensional arrays. That builder is appended to the final component type in the base
+     * case.
+     *
+     * @param e same as above, but can become null if scene-lib did not fill in the inner types,
+     *     which happens when they do not have annotations
+     * @param arrayTypes the array parts of the array type (i.e. the parts after the component type
+     *     in the string representation). Must contain at least one '['.
+     * @param componentType the component type of the array
+     * @param result the string builder containing the array types seen so far
+     * @return the formatted string, without a trailing space
+     */
+    private static String formatArrayTypeImpl(
+            @Nullable ATypeElement e,
+            String arrayTypes,
+            String componentType,
+            StringBuilder result) {
+        // append the next type:
+        String nextArrayType = arrayTypes.substring(0, arrayTypes.indexOf(']') + 1);
+        String remainingArrayTypes = arrayTypes.substring(arrayTypes.indexOf(']') + 1);
+        // do not append inferred annotations if there was one in the source code
+        if (nextArrayType.contains("@")) {
+            result.append(nextArrayType);
+        } else {
+            if (e != null) {
+                result.append(formatAnnotations(e.tlAnnotationsHere));
+            }
+            result.append(nextArrayType);
+        }
+        result.append(" ");
+
+        // Check if there are any more array levels. If so, recurse; otherwise, append the component
+        // and return.
+        if ("".equals(remainingArrayTypes)) {
+            ATypeElement component = getNextArrayLevel(e);
+            return formatType(componentType, component) + result.toString();
+        } else {
+            return formatArrayTypeImpl(
+                    getNextArrayLevel(e), remainingArrayTypes, componentType, result);
+        }
+    }
+
+    /**
+     * Gets the outermost array level (or the component if not an array) from the given type
+     * element, or null if scene-lib is not storing any more information about this array (for
+     * example, when the component type is unannotated).
+     *
+     * @param e the array type element; can be null
+     * @return the next level of the array, if scene-lib stores information on it. null if the input
+     *     is null or scene-lib is not storing more information.
+     */
+    private static @Nullable ATypeElement getNextArrayLevel(@Nullable ATypeElement e) {
+        if (e == null) {
+            return null;
+        }
+
+        for (Map.Entry<InnerTypeLocation, ATypeElement> ite : e.innerTypes.entrySet()) {
+            InnerTypeLocation loc = ite.getKey();
+            if (loc.location.contains(TypePathEntry.ARRAY)) {
```

Comment:
What if the component type contains an array, as in
`List<String[]> []`
?  Could some wrong annotations be output?

Reply:
I don't understand how the scene-lib representation works well enough to answer your question. This code was copied from `IndexFileWriter`, though, so I will just move this method there.

Reply:
Actually, I can't find that code in `IndexFileWriter` anymore; either it was removed (doubtful) or I've modified it beyond recognition. I'll add a test.

Reply:
I wrote a test for this and both the stub-based WPI and JAIF-based WPI failed it, because they don't infer type arguments. I'll commit a weaker version of my test with the failing parts commented out.